# <center><u>**Mini-Project**</u></center>

In [15]:
import numpy as np
import pandas as pd

In [5]:
driv_df = pd.read_csv("drivPoints.txt")

In [7]:
display(driv_df)

fileName  subject  imgNum  label  ang   xF   yF   wF   hF  \
0    20130529_01_Driv_001_f         1       1      2    0  292  209  100  112   
1    20130529_01_Driv_002_f         1       2      2    0  286  200  109  128   
2    20130529_01_Driv_003_f         1       3      2    0  290  204  105  121   
3    20130529_01_Driv_004_f         1       4      2    0  287  202  112  118   
4    20130529_01_Driv_005_f         1       5      2    0  290  193  104  119   
..                       ...      ...     ...    ...  ...  ...  ...  ...  ...   
601  20130530_04_Driv_086_f         4      86      2  -15  278  183  128  141   
602  20130530_04_Driv_087_lr        4      87      1  -30  268  186  128  134   
603  20130530_04_Driv_088_lr        4      88      1  -30  264  187  127  131   
604  20130530_04_Driv_089_f         4      89      2  -15  264  175  143  136   
605  20130530_04_Driv_090_f         4      90      2    0  266  170  141  139   

     xRE  yRE  xLE  yLE   xN   yN  xRM  yRM  xLM  yLM  
0    323  232  367  231  353  254  332  278  361  278  
1    324  235  366  235  353  258  333  281  361  281  
2    325  240  367  239  351  260  334  282  362  282  
3    325  230  369  230  353  253  335  274  362  275  
4    325  224  366  225  353  244  333  268  363  268  
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  
601  307  218  354  210  330  247  324  273  356  266  
602  296  222  344  212  319  247  316  274  347  269  
603  287  220  334  211  304  247  305  272  337  270  
604  295  207  345  200  320  234  314  261  351  251  
605  303  206  354  198  331  229  319  255  362  247  

[606 rows x 19 columns]

***

## Dataset Description

### *Input*

The DrivFace database contains images sequences of subjects while driving in real scenarios. 
It is composed of 606 samples of 640×480 pixels each, acquired over different days from 4 drivers.

In [18]:
print("Number of Different Drivers:", len(driv_df['subject'].unique()))

Number of Different Drivers: 4


### *Label*

A set of labels assigning each image into 3 possible gaze direction classes are given.
* The first class is the “looking-right” class and contains the head angles between -45º and -30º.
* The second one is the “frontal” class and contains the head angles between -15º and 15º.
* The last one is the “looking-left” class and contains the head angles between 30º and 45º.

### *Features*

* fileName is the image's name into DrivImages.zip
* subject = [1:4]
* imgNum  = int
* label   = [1/2/3] (head pose class that corresponding to [lr/f/lf], respectively) 
* ang     = [-45, -30/ -15 0 15/ 30 15] (head pose angle)
* [xF yF wF hF] = face position 
* [xRE yRE]     = rigth eye position 
* [xLE yLE]      = left eye position 
* [xN yN]       = Nose position 
* [xRM yRM]     = rigth corner of mouth
* [xLM yLM]     = left corner of mouth   

***

In [24]:
display(driv_df.corr()['label'])

subject    0.031872
imgNum     0.050105
label      1.000000
ang        0.785522
xF         0.408662
yF        -0.049579
wF        -0.007389
hF         0.086531
xRE        0.533790
yRE       -0.058692
xLE        0.485336
yLE       -0.039557
xN         0.588441
yN        -0.061158
xRM        0.567880
yRM       -0.056930
xLM        0.491974
yLM       -0.037165
Name: label, dtype: float64

**Note:** From above cor-relation table b/w head direction and other features it is clear that head position is weakly cor-related to the y-cordinates of any face feature, which makes sense, and it is also weakly cor-related to the driver's face width and height, hence we can simply drop these colums from our dataset.

#### Features that the model will use to predict the label

In [31]:
features = ['ang', 'xF', 'xRE', 'xLE', 'xN', 'xRM', 'xLM']

In [33]:
driv_df_2 = driv_df[features]
driv_df_2['label'] = driv_df['label']

<ipython-input-33-169a14e28c55>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  driv_df_2['label'] = driv_df['label']


In [30]:
display(driv_df_2)

ang   xF  xRE  xLE   xN  xRM  xLM
0      0  292  323  367  353  332  361
1      0  286  324  366  353  333  361
2      0  290  325  367  351  334  362
3      0  287  325  369  353  335  362
4      0  290  325  366  353  333  363
..   ...  ...  ...  ...  ...  ...  ...
601  -15  278  307  354  330  324  356
602  -30  268  296  344  319  316  347
603  -30  264  287  334  304  305  337
604  -15  264  295  345  320  314  351
605    0  266  303  354  331  319  362

[606 rows x 7 columns]